### PR TITLE
Mes 1713 use iam auth for poller rds access

### PR DESCRIPTION
**Pull request by JamieKellyBJSS**
MES 1713 

- Use mysql2 instead of mysql as mysql cannot handle IAM RDS auth
- Use IAM RDS auth instead of secrets in secrets manager

## Pull Request checklist

- [x] [WIP] tag removed from PR title
- [x] PR has an understandable description

## Git feature branch checklist

- [x] branch name comply with our branching strategy
- [x] git branch contains relevant JIRA ticket number
- [x] branch rebased against the latest develop

## Sign off process checklist

- [x] Code has been tested manually
- [ ] Tested by QA
- [ ] PO's approval
- [ ] Reviewers selected in Github
- [x] PR link added to JIRA

### CODE REVIEW

File: src/common/framework/config/__tests__/config-helpers.spec.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import { Mock, It, Times } from 'typemoq';
```
Comment (JamieKellyBJSS) [inline src/common/framework/config/__tests__/config-helpers.spec.ts]: These are not used, delete.

File: src/common/framework/config/__tests__/config-helpers.spec.ts [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+import { tryFetchRdsAccessToken, generateSignerOptions } from '../config-helpers';
```
Comment (JamieKellyBJSS) [inline src/common/framework/config/__tests__/config-helpers.spec.ts]: We don't use tryFetchRdsAccessToken

File: src/common/framework/config/config-helpers.ts [hunk 3]
```diff
@@ -15,54 +15,58 @@ export const throwIfNotPresent = (value: string | null | undefined, configKey: s
...
+  if (!iamRdsConfigValid(hostname, username)) {
     const envvar = process.env[fallbackEnvvar];
     if (!envvar) {
       throw new Error(`No value for fallback envvar ${fallbackEnvvar} for config`);
     }
     return envvar;
   }
 
-  await tryBootstrapSecretsManager();
+  throwIfNotPresent(
```
Comment (JamieKellyBJSS) [inline src/common/framework/config/config-helpers.ts]: Are these both not irrelevant given the above checks?

File: src/common/framework/config/config-helpers.ts [hunk 4]
```diff
@@ -15,54 +15,58 @@ export const throwIfNotPresent = (value: string | null | undefined, configKey: s
...
+    hostname,
+    'tarsReplicateDatabaseHostname',
```
Comment (johnmorleybjss) [inline src/common/framework/config/config-helpers.ts]: These args aren't too long, why not inline them?

File: src/common/framework/config/config-helpers.ts [hunk 5]
```diff
@@ -15,54 +15,58 @@ export const throwIfNotPresent = (value: string | null | undefined, configKey: s
...
+    username,
+    'tarsReplicaDatabaseUsername',
+  );
+
+  const signer = new RDS.Signer();
+  const signerOptions = generateSignerOptions(hostname, username);
+  return new Promise((resolve, reject) => {
+    signer.getAuthToken(signerOptions, (err, token) => {
+      if (err) {
+        throw new Error(`Generating an auth token failed`);
```
Comment (johnmorleybjss) [inline src/common/framework/config/config-helpers.ts]: I think we should attach `err` into the `Error` here for more visibility of root cause in the logs.

File: src/common/framework/config/config-helpers.ts [hunk 6]
```diff
@@ -15,54 +14,45 @@ export const throwIfNotPresent = (value: string | null | undefined, configKey: s
...
 
-  const valueFromSecretsManager = get(secretsManagerConfig, <string>secretKey);
-  if (!valueFromSecretsManager) {
-    throw new Error(`The key ${secretKey} was not found in the ASM config`);
-  }
-  return valueFromSecretsManager;
+  const signer = new RDS.Signer();
+  const signerOptions = generateSignerOptions(hostname, username);
+  return new Promise((resolve, reject) => {
+    signer.getAuthToken(signerOptions, (err, token) => {
```
Comment (ammarhaiderbjss) [inline src/common/framework/config/config-helpers.ts]: Fine as it is but for future reference all aws sdk methods for node come with a `.promise()` to convert it to a promise

File: src/functions/pollJournals/framework/repo/mysql/pool.ts [hunk 7]
```diff
@@ -8,6 +8,12 @@ export const createConnectionPool = (): mysql.Pool => {
     database: configuration.tarsReplicaDatabaseName,
     user: configuration.tarsReplicaDatabaseUsername,
     password: configuration.tarsReplicaDatabasePassword,
+    ssl: process.env.TESTING_MODE ? null : 'Amazon RDS',
+    authSwitchHandler(data, cb) {
```
Comment (ammarhaiderbjss) [inline src/functions/pollJournals/framework/repo/mysql/pool.ts]: What's this for?

File: src/common/framework/config/config-helpers.ts (hunk 6 above)
Comment (JamieKellyBJSS) [reply src/common/framework/config/config-helpers.ts]: This one doesn't unfortunately!

File: src/functions/pollJournals/framework/repo/mysql/pool.ts (hunk 7 above)
Comment (JamieKellyBJSS) [reply src/functions/pollJournals/framework/repo/mysql/pool.ts]: This is the configuration that enables the IAM RDS Auth to work on the connection, as seen in the example here: https://cloudonaut.io/passwordless-database-authentication-for-aws-lambda/

**Pull request merged**